tapdb/assets_store.go

```diff
@@ -320,6 +322,17 @@ type AssetStore struct {
 	// eventDistributor is an event distributor that will be used to notify
 	// subscribers about new proofs that are added to the archiver.
 	eventDistributor *fn.EventDistributor[proof.Blob]
+
+	// coinLock is a read/write mutex that is used to ensure that only one
+	// goroutine is attempting to call any coin selection related methods at
+	// any time. This is necessary as some of the calls to the store (e.g.
+	// ListEligibleCoins -> LeaseCoin) are called after each other and
+	// cannot be placed within the same database transaction. So calls to
+	// those methods must hold this coin lock. We use this in conjunction
+	// with an atomic  boolean, so we can check if the lock is held without
+	// having to acquire the mutex.
+	coinLock     sync.Mutex
```

Comment:
I think this should live one level higher in the porter. So here: https://github.com/lightninglabs/taproot-assets/blob/edf59c62d5e4acd407d2675414e5d86e78eaf47b/tapfreighter/wallet.go#L143. 

In the past we've had some hard to track down bugs as a result of inserting user-space DB level mutexes. This way we also make minimal assumptions re concurrency control beyond what's provided by the DB.